### ISSUE TITLE
manual_map false positive

### ISSUE THREAD

**Issue by Abcight** [rust-lang/rust-clippy#10178]
### Summary

Clippy suggests replacing ``match`` on ``Option`` with ``map`` without regard to ownership rules.

### Lint Name

manual_map

### Reproducer

Consider this code:

```rust
let mut vector = vec![1, 2, 3];
let first = match vector.get(0) {
    Some(_) => Some(vector.swap_remove(0)),
    None => None
};
```

Clippy will suggest replacing match with map:

```rust
let mut vector = vec![1, 2, 3];
let first = vector.get(0).map(|_| vector.swap_remove(0));
```

Which does not compile.

### Version

```text
rustc 1.66.0 (69f9c33d7 2022-12-12)
binary: rustc
commit-hash: 69f9c33d71c871fc16ac445211281c6e7a340943
commit-date: 2022-12-12
host: x86_64-pc-windows-msvc
release: 1.66.0
LLVM version: 15.0.2
```

**Comment by Jarcho** [comment]
Same issue as #8346. This specific pattern can be fixed without integrating the borrow checker so it's worth keeping open as a separate issue.

**Comment by TimonPost** [comment]
This code also gives an error:

```rust
use std::sync::Arc;

fn main() {
    gives_clippy_warning();   
    clippy_suggestion_gives_error();
}

fn my_function(foo: &String) -> String {
    foo.to_string()
}
        
fn gives_clippy_warning() {
    let closure = |foo| my_function(foo);

    let value_a = "a".to_string();
    let value_b = Arc::new("b".to_string());
    
    // clippy: manual_map warning
    let _result = if 1 == 2 {
        Some(closure(&value_a))
    } else if let Some(value_b) = Some(value_b) {
        Some(closure(&value_b))
    } else {
        None
    };
}

fn clippy_suggestion_gives_error() {
    let closure = |foo| my_function(foo);

    let value_a = "a".to_string();
    let value_b = Arc::new("b".to_string());

    // Clippy suggestion but doesnt compile as value_b does not live long enough
    let _result = if 1 == 2 {
        Some(closure(&value_a))
    } else {
        Some(value_b).map(|value_b| closure(&value_b))
    };
}
```
